### PR TITLE
Views support (#52)

### PR DESCRIPTION
**Pull request by sergey-netdev**
Added extra condition to generate FKs only for user tables. #52

### CODE REVIEW

File: EfEnumToLookupTests/Tests/TestStuff.cs [hunk 1]
```diff
@@ -42,11 +48,41 @@ public void DoesStuff()
...
+
+				ViewRabbit actual = context.ViewRabbits.First();
+				Assert.AreEqual("Roger", actual.Name);
+				Assert.AreEqual(Ears.Pointy, actual.TehEars);
+				Assert.AreEqual(Fur.Brown, actual.BodyFur);
+			}
+		}
+
+		/// <summary>Context containing a view</summary>
+		public class ViewDbContext : DbContext
```
Comment (timabell) [inline EfEnumToLookupTests/Tests/TestStuff.cs]: this should be in its own file in the db/ folder

File: EfEnumToLookupTests/Tests/TestStuff.cs [hunk 2]
```diff
@@ -42,11 +48,41 @@ public void DoesStuff()
...
+		{
+			public ViewDbContext(string nameOrConnectionString) : base(nameOrConnectionString) { } // to override db naming convention to use existing db created by MagicContext
+
+			public DbSet<ViewRabbit> ViewRabbits { get; set; }
+		}
+
+		public class ViewRabbit : Rabbit { }
```
Comment (timabell) [inline EfEnumToLookupTests/Tests/TestStuff.cs]: This should go in the model/ folder

File: EfEnumToLookup/LookupGenerator/SqlServerHandler.cs [hunk 3]
```diff
@@ -100,7 +100,7 @@ private string AddForeignKeys(IEnumerable<EnumReference> refs)
 				var fkName = string.Format("FK_{0}_{1}", enumReference.ReferencingTable, enumReference.ReferencingField);
 
 				sql.AppendFormat(
-					" IF OBJECT_ID('{0}', 'F') IS NULL ALTER TABLE [{1}] ADD CONSTRAINT {0} FOREIGN KEY ([{2}]) REFERENCES [{3}] (Id);\r\n",
+					" IF OBJECT_ID('{0}', 'F') IS NULL AND OBJECT_ID('{1}', 'U') IS NOT NULL ALTER TABLE [{1}] ADD CONSTRAINT {0} FOREIGN KEY ([{2}]) REFERENCES [{3}] (Id);\r\n",
```
Comment (timabell) [inline EfEnumToLookup/LookupGenerator/SqlServerHandler.cs]: Neat solution!

It bothers me a little that the model we have in memory still thinks there's a fk, which might make future work on the library more challenging. Having said that I guess EF is unaware it's a view so we're stuffed. I'll take this solution as I don't have any better ideas :-)

File: EfEnumToLookupTests/Tests/TestStuff.cs [hunk 4]
```diff
@@ -42,11 +48,41 @@ public void DoesStuff()
 				var actual = context.PeskyWabbits.First();
 				Assert.AreEqual("Roger", actual.Name);
 				Assert.AreEqual(Ears.Pointy, actual.TehEars);
+				Assert.AreEqual(Fur.Brown, actual.BodyFur);
 				Assert.AreEqual(1, context.PeskyWabbits.Count()); // spot unwanted re-use of db
 			}
 		}
 
 		[Test]
+		public void DoesStuffWithViews()
```
Comment (timabell) [inline EfEnumToLookupTests/Tests/TestStuff.cs]: thanks for expanding the tests to cover this, coverage is important to me on this library

**Comment by timabell** [comment]
Nice work!

Just a couple of trivial comments then I expect this can go in. I'll probably give it a poke on my own machine first just to be sure. Seeing as this fixes a bug in a new use case I think this can probably go out in the next point release rather than waiting for a minor or major release.

**Comment by timabell** [comment]
Another thought, it would be worth adding something to the readme explaining the that although ef can use views, the fks won't be created and that you'll be on your own for linking any tables ef doesn't know about.

**Comment by sergey-netdev** [comment]
Good idea! I'll add a few lines.

**Comment by timabell** [comment]
I'm going to test it, squash the first two commits, and probably add my own take to the readme, then merge. Unless you want to do do the squash?

**Comment by sergey-netdev** [comment]
I'm still a bit new to github so I can only guess what the squash is (rebase maybe :)

**Comment by timabell** [comment]
yeah, nothing to do with github, good old fashioned `rebase --interactive`

**Comment by timabell** [comment]
I've had a proper play with this now and in looking closer at the test you've provided I can see now that it's very tightly tied to the existing tests, reusing the same database. I'm not happy with this as it makes the test suite much harder to maintain in the long run. I had a go at disentangling them from each other, but it seems that ef really doesn't expect you to use views, so I quickly ran into serious problems. I'm guessing you aren't using the automatic database generation that the tests use as that seems to make it impossible to do cleanly.

As it stands, I'd consider taking a patch to just the generated sql + the readme, without the additional test as I think views are an edge case and this is the first report I've had relating to them. So long as it doesn't break the existing tests (which it currently doesn't) then I see no harm.

To add test coverage, I think the best approach might be to create a completely unrelated dbcontext, model and test class which matches more closely your real use of the library, including whatever method you're using to build your db (migrations? raw sql scripting?).

Up to you what you want to do now, feel free to open a fresh PR from a new branch (make sure you base it on the v1 branch this time) with a more minimal patch, or an alternate approach to the tests (which would be appreciated but isn't required).

**Comment by sergey-netdev** [comment]
As we fight performance issues we lean toward db first approach. We still use migrations but the number of SPs, views and functions grows so I expect we'll switch to completely db first in a few months.

You're right that the test is dependent on the existing context but so are views on existing tables (which are often just a bunch of heavy joins and the related entity classes are often inherited). That's why it took me some time to figure out how to add the test. Probably adding a script to generate unrelated dbcontext is a good idea to simulate db first approach but the current test definitely reflects the scenario we got.

**Comment by timabell** [comment]
Are you using those `Up()` / `Down()` methods that EF provides or some other migration system?

P.s. take a look at ready-roll, it's very good.

**Comment by timabell** [comment]
It seems you can't mix the automatic db generation with migrations easily.

http://stackoverflow.com/questions/20862807/mapping-database-views-to-ef-5-0-code-first-w-migrations

**Comment by sergey-netdev** [comment]
Yes, we use Up/Down methods and it's really painful to mix arbitrary sql text with native EF migrations.
